docs: dockerhub webhook (#4420)
akuity/kargo: docs/docs/50-user-guide/60-reference-docs/80-webhook-receivers/dockerhub/index.md

---
sidebar_label: Docker Hub
description: How to configure Docker Hub webhooks with Kargo for automated artifact discovery.
---

# Docker Hub Webhook Receiver

The Docker Hub Webhook Receiver responds to `push` events originating from Docker Hub repositories.

In response to a `push` event, the receiver "refreshes" `Warehouse` resources subscribed to the corresponding Docker Hub repository.

:::info
"Refreshing" a `Warehouse` means enqueuing it for immediate reconciliation by the Kargo controller, which will attempt to discover new artifacts from all subscribed repositories.
:::

:::note

> Unlike GitHub, Docker Hub does **not** sign webhook payloads. Kargo uses a static token for basic validation.

Reviewer: Referring to GitHub here does not make much sense IMHO, as the user is interested in configuring DockerHub (and may not be aware of GitHub _at all_).

:::

## Configuring the Receiver

To enable webhook support for Docker Hub, you must configure a Kubernetes `Secret` and reference it in your `ProjectConfig`.

> The `Secret` must include a `dockerhub-secret` key under `stringData`. This value acts as a token for lightweight validation of incoming webhook requests.

Reviewer: ```suggestion
The `Secret` must include a `secret` key under `stringData`. This value acts as a token for lightweight validation of incoming webhook requests.
```

Reviewer 2: I think there are much more accurate ways to describe this than a "token for lightweight validation." We could pretty easily explain that it's used in generating a hard-to-guess URL for the receiver. Search existing docs for wording you can reuse. It's mentioned in places like the working with Warehouses page.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: dh-wh-secret
  namespace: kargo-demo
stringData:

>   dockerhub-secret: <your-static-token-here>

Reviewer: ```suggestion
  secret: <your-static-token-here>
```

````

Next, reference this `Secret` in your `ProjectConfig`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers:
    - name: dockerhub-receiver
      dockerhub:
        secretRef:
          name: dh-wh-secret
```

## Retrieving the Receiver's URL

After applying the `ProjectConfig`, Kargo will generate a unique, hard-to-guess URL for the Docker Hub webhook receiver.

You can retrieve this URL using the following command:

```shell
kubectl get projectconfigs kargo-demo \
  -n kargo-demo \
  -o=jsonpath='{.status.webhookReceivers}'
```

> Copy the URL associated with your Docker Hub webhook receiver. You'll use this in the next step.

Reviewer: ```suggestion
```

This is a reference document, and sections in this document are typically standalone. Because of this, instructions to "copy" something should not be made.

Instead, within "registering with Docker Hub", you would reference back to this section to instruct to user in how to obtain the value.

## Registering with Docker Hub

To configure a Docker Hub repository to send events to the webhook receiver:

1. Navigate to your Docker Hub repository and select the <Hlt>Webhooks</Hlt> tab.

   ![Webhooks Tab](./img/webhooks-tab.png "Webhooks Tab")

1. In the <Hlt>Webhooks</Hlt> form:

   ![New Webhook](./img/new-webhook.png "New Webhook Form")

   1. Provide a name for the webhook.

>    1. Set <Hlt>Webhook URL</Hlt> to the value you
>       [retrieved earlier](#retrieving-the-receivers-url).

Reviewer: ```suggestion
   1. Set <Hlt>Webhook URL</Hlt> to the
      [receiver URL](#retrieving-the-receivers-url).
```

   1. Click <Hlt>+</Hlt> to create webhook.

      ![Create Webhook](./img/create-webhook.png "Create Webhook Button")

## Verifying Webhook Delivery

> After pushing a new image to your Docker Hub repository:

Reviewer: I would rephrase this into something like:

```suggestion
To verify your webhook is correctly registered and working as expected, after pushing a new image to your Docker Hub repository:
```

1. Return to the <Hlt>Webhooks</Hlt> tab of your repository.

1. In the <Hlt>Current Webhooks</Hlt> section, hover over your webhook, select the *menu options* icon, and click <Hlt>View History</Hlt>.

   ![View Webhook History](./img/view-history.png "View History")

1. Check the delivery log to confirm a successful webhook request.

   ![Delivery Detail](./img/delivery-detail.png "Webhook Delivery Detail")

   Ensure the response status is `200 OK`.

> When these steps are complete, Kargo will automatically refresh the corresponding `Warehouse` and initiate artifact discovery from your updated Docker Hub repository.

Reviewer: I think this should be put under "Registering with Docker Hub".

---
sidebar_label: Docker Hub
description: How to configure Docker Hub webhooks with Kargo for automated artifact discovery.
---

# Docker Hub Webhook Receiver

> The Docker Hub Webhook Receiver responds when container images or charts are

Reviewer 2: The precision of this statement could be better. It responds to _notifications_ of images or charts having been pushed. The role of the webhook request is an important detail. Without that, this seems like magic.

pushed to Docker Hub repositories.

When this happens, the receiver "refreshes" `Warehouse` resources subscribed to
the corresponding Docker Hub repository.

:::info
"Refreshing" a `Warehouse` means enqueuing it for immediate reconciliation by
the Kargo controller, which will attempt to discover new artifacts from all
subscribed repositories.
:::

## Configuring the Receiver

To enable webhook support for Docker Hub, you must configure a Kubernetes
`Secret` and reference it in your `ProjectConfig`.

The `Secret` must include a `secret` key in its data map. This value is used to
generate a unique, hard-to-guess URL for the webhook receiver, providing basic
protection against unauthorized requests. Because Docker Hub webhook payloads
are **not** signed, Kargo uses this static token for basic validation.

:::note

> The following command is suggested for generating a complex shared secret and

Reviewer 2: wrt Docker Hub, "shared secret" is not correct terminology.

encoding it for use in the `data` field:

```shell
openssl rand -base64 48 | tr -d '=+/' | head -c 32 | base64
```

:::

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: dh-wh-secret
  namespace: kargo-demo
data:

>   secret: <your-base64-encoded-token-here>

Reviewer 2: The word "token" wasn't used anywhere up to this point. It's been referred to as a secret and a shared secret (incorrect). I know it must seem silly, but if we're not consistent with our language here it _will_ confuse someone.

---
sidebar_label: Docker Hub
description: How to configure Docker Hub webhooks with Kargo for automated artifact discovery.
---

# Docker Hub Webhook Receiver

The Docker Hub Webhook Receiver responds to `push` events originating from
Docker Hub repositories.

In response to a `push` event, the receiver "refreshes" `Warehouse` resources

> subscribed to the corresponding Docker Hub repository.

Reviewer 2: In contrast to some other platforms, I don't see any place in the Docker Hub docs where the webhooks are associated with any specific event names/types. So the choice to monospace `push` is incorrect. I would probably not even call it a "push event" at all. I would probably use more generic wording like "when container images or charts are pushed to a Docker Hub repository" to describe the event.

:::info
"Refreshing" a `Warehouse` means enqueuing it for immediate reconciliation by
the Kargo controller, which will attempt to discover new artifacts from all
subscribed repositories.
:::

:::note
Docker Hub webhook payloads are **not** signed. Kargo uses a static token for
basic validation.
:::

## Configuring the Receiver

To enable webhook support for Docker Hub, you must configure a Kubernetes
`Secret` and reference it in your `ProjectConfig`.

The `Secret` must include a `secret` key under `stringData`. This value is used
to generate a unique, hard-to-guess URL for the webhook receiver, providing

> basic protection against unauthorized requests.

Reviewer 2: It also wouldn't be terrible to simply use `data` and account for base 64 encoding in the suggested command for generating the secret. (In which case, the same improvement could probably be made to the GitHub page.)

fwiw, you can also have fun with this. If you look at the Working with Warehouses page and decode my Secrets, they're easter eggs.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: dh-wh-secret
  namespace: kargo-demo
stringData:
  secret: <your-static-token-here>
```

Next, reference this `Secret` in your `ProjectConfig`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers:
    - name: dockerhub-receiver
      dockerhub:
        secretRef:
          name: dh-wh-secret

> ```

Reviewer 2: Just put these together in one chunk of YAML. It's a tidier, more comprehensive view of what's being done. _This is how the GitHub page is doing it._

```

## Registering with Docker Hub

To configure a Docker Hub repository to send events to the webhook receiver:

1. Navigate to your Docker Hub repository and select the <Hlt>Webhooks</Hlt> tab.

   ![Webhooks Tab](./img/webhooks-tab.png "Webhooks Tab")

1. In the <Hlt>Webhooks</Hlt> form:

   ![New Webhook](./img/new-webhook.png "New Webhook Form")

   1. Provide a name for the webhook.

   1. Set <Hlt>Webhook URL</Hlt> to the
      [receiver URL](#retrieving-the-receivers-url).

   1. Click <Hlt>+</Hlt> to create webhook.

      ![Create Webhook](./img/create-webhook.png "Create Webhook Button")

## Verifying Webhook Delivery

To verify your webhook is correctly registered and working as expected, after
pushing a new image to your Docker Hub repository:

1. Return to the <Hlt>Webhooks</Hlt> tab of your repository.

1. In the <Hlt>Current Webhooks</Hlt> section, hover over your webhook, select
the *menu options* icon, and click <Hlt>View History</Hlt>.

   ![View Webhook History](./img/view-history.png "View History")

1. Check the delivery log to confirm a successful webhook request.

   ![Delivery Detail](./img/delivery-detail.png "Webhook Delivery Detail")

>    Ensure the response status is `200 OK`.

Reviewer 2: Fix formatting please.

   1. Provide a name for the webhook.

   1. Set <Hlt>Webhook URL</Hlt> to the
      [receiver URL](#retrieving-the-receivers-url).

   1. Click <Hlt>+</Hlt> to create webhook.

      ![Create Webhook](./img/create-webhook.png "Create Webhook Button")

> ## Verifying Webhook Delivery

Reviewer: Looking more closely at the GitHub example, this whole section seems self-invented and not in line with the format of the other document.
